src/VectorizeLoops.cpp

```diff
@@ -10,23 +10,209 @@
 #include "IREquality.h"
 #include "ExprUsesVar.h"
 #include "Solve.h"
+#include "Simplify.h"
+#include "CSE.h"
 
 namespace Halide {
 namespace Internal {
 
 using std::string;
 using std::vector;
+using std::pair;
+
+// For a given var, replace expressions like shuffle_vector(var, 4)
+// with var.lane.4
+class ReplaceShuffleVectors : public IRMutator {
+    string var;
+
+    using IRMutator::visit;
+
+    void visit(const Call *op) {
+        const Variable *v;
+        const int64_t *i;
+        if (op->is_intrinsic(Call::shuffle_vector) &&
+            op->args.size() == 2 &&
+            (v = op->args[0].as<Variable>()) &&
+            v->name == var &&
+            (i = as_const_int(op->args[1]))) {
+            expr = Variable::make(op->type, var + ".lane." + std::to_string(*i));
+        } else {
+            IRMutator::visit(op);
+        }
+    }
+public:
+    ReplaceShuffleVectors(const string &v) : var(v) {}
+};
+
+
+
+/** Find the exact max and min lanes of a vector expression. Not
+ * conservative like bounds_of_expr, but uses similar rules for some
+ * common node types where it can be exact. Assumes any vector
+ * variables defined externally also have .min_lane and .max_lane
+ * versions in scope. */
+Interval bounds_of_lanes(Expr e) {
+    if (const Add *add = e.as<Add>()) {
+        if (const Broadcast *b = add->b.as<Broadcast>()) {
+            Interval ia = bounds_of_lanes(add->a);
```

Comment:
Would it improve readability to pull these two lines out into a helper function / lambda?